tests/test_xlora.py

```diff
@@ -0,0 +1,172 @@
+
+import os
+
+import torch
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+from peft import LoraConfig, PeftType, TaskType, XLoraConfig, get_peft_model
+
+
+model_id = "facebook/opt-125m"
+tokenizer = AutoTokenizer.from_pretrained(model_id, trust_remote_code=True, device_map="cuda:0")
+tmp_path = "/tmp/xlora_test"
+os.makedirs(tmp_path, exist_ok=True)
+
+for i in range(1, 9):
+    torch.manual_seed(i)
+    lora_config = LoraConfig(task_type="CAUSAL_LM", init_lora_weights=False)
+    model = AutoModelForCausalLM.from_pretrained(model_id)
+    peft_model = get_peft_model(model, lora_config)
+    peft_model.save_pretrained(f"{tmp_path}/checkpoint-{i}")
+    print(f"finished {i} of 8")
+
+
+class TestXlora:
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+    def test_functional(self):
+        tokenizer = AutoTokenizer.from_pretrained(model_id, trust_remote_code=True, device_map="cuda:0")
+
+        model = AutoModelForCausalLM.from_pretrained(model_id)
+        model.config.use_cache = False
+        adapters = {str(i): f"{tmp_path}/checkpoint-{i}" for i in range(1, 9)}
+
+        peft_config = XLoraConfig(
+            task_type=TaskType.CAUSAL_LM,
+            peft_type=PeftType.XLORA,
+            hidden_size=model.config.hidden_size,
+            xlora_depth=8,
+            adapters=adapters,
+        )
+        model = get_peft_model(model, peft_config).to("cuda")
+
+        model.enable_scalings_logging()
+        inputs = tokenizer.encode("Python is a", add_special_tokens=False, return_tensors="pt")
+        outputs = model.generate(
+            input_ids=inputs.to("cuda"),
+            max_new_tokens=32,
+        )
+        text = tokenizer.batch_decode(outputs[: inputs.shape[1] :].detach().cpu().numpy(), skip_special_tokens=True)
+        print(text[0])
+
+    def test_scalings_logging_methods(self):
+        tokenizer = AutoTokenizer.from_pretrained(model_id, trust_remote_code=True, device_map="cuda:0")
+
+        model = AutoModelForCausalLM.from_pretrained(model_id)
+        model.config.use_cache = False
+        adapters = {str(i): f"{tmp_path}/checkpoint-{i}" for i in range(1, 9)}
+
+        peft_config = XLoraConfig(
+            task_type=TaskType.CAUSAL_LM,
+            peft_type=PeftType.XLORA,
+            hidden_size=model.config.hidden_size,
+            xlora_depth=8,
+            adapters=adapters,
+        )
+        model = get_peft_model(model, peft_config).to("cuda")
+
+        model.enable_scalings_logging()
+
+        inputs = tokenizer.encode("Python is a", add_special_tokens=False, return_tensors="pt")
+        outputs = model.generate(
+            input_ids=inputs.to("cuda"),
+            max_new_tokens=32,
+        )
+        text = tokenizer.batch_decode(outputs[: inputs.shape[1] :].detach().cpu().numpy(), skip_special_tokens=True)
+        print(text[0])
+
+        _ = model.get_latest_scalings()
+        assert 32 >= len(model.get_scalings_log()) > 0
```

Comment:
With these "magic numbers" like 32, it's always nice to add a comment to explain where they come from.